File: hackerearth/Monk and the Islands.cpp
#include<bits/stdc++.h>
using namespace std;

int main()
{
    long loop, test, node, edge, a, b, dist[100000];
    vector<long>V[100000], cost[100000];
    queue<long>Q;
    scanf("%ld", &test);
    for(loop=1 ; loop<=test ; loop++)
    {
        scanf("%ld%ld", &node, &edge);
        while(edge--)
        {
            scanf("%ld%ld", &a, &b);
            V[a].push_back(b);
            V[b].push_back(a);

            cost[a].push_back(b);
            cost[b].push_back(a);
        }

        memset(dist, 0, sizeof dist);

        for(int i=2 ; i<=node ; i++)
            dist[i] = 99999999;
        
        Q.push(1);
        while(!Q.empty())
        {
            a = Q.front();
            Q.pop()
            for(int i=0 ; i<V[a].size() ; i++)
                if(dist[ V[a][i] ] > dist[a]+1)
                    dist[ V[a][i] ] = dist[a]+1;
        }

        printf("%ld\n", dist[node-1]);

        for(int i=0 ; i<=node ; i++)
        {
            V[i].clear();
            cost[i].clear();
        }
    }
}

File: hackerearth/Monk and the Islands.py
for _ in range(0, int(input())):
    aa = input()
    [node, edge] = [int(s) for s in aa.split() if s.isdigit()]
    
    V = []
    dist = []
    Q = []

    for i in range(0, node+1):
        V.append([])
        dist.append(999999)
    
    while(edge):
        aaa = input()
        [a, b] = [int(s) for s in aaa.split() if s.isdigit()]
        V[a].append(b)
        V[b].append(a)
        edge = edge - 1

    dist[1] = 0
    Q.append(1)
    
    while(len(Q)):
        a = Q[0]
        Q = Q[1:]

        for i in V[a]:
            if dist[i] > dist[a] + 1: 
                dist[i] = dist[a] + 1
                Q.append(i)
    print(dist[node])